In [372]:
# Importing libraries
import pandas as pd
from os import listdir
from os.path import isfile, join
import xgboost as xgb

# Reading the data
data_path = 'Data/'
log = pd.read_csv(data_path + 'LOG_WEB_201708.csv', sep=';')
clients = pd.read_csv(data_path + 'clients.csv', sep = ';')
cmd = pd.read_csv(data_path + 'CMD.csv', sep = ';')
cmd = cmd[cmd['ORDER_DATE'] >= '2017-08-01']
table = pd.read_csv(data_path + 'TABLE_CONVERSION.csv', sep = ';')

# Remove unamed column
log = log.loc[:, ~log.columns.str.contains('^Unnamed')]
clients = clients.loc[:, ~clients.columns.str.contains('^Unnamed')]
cmd = cmd.loc[:, ~cmd.columns.str.contains('^Unnamed')]
table = table.loc[:, ~table.columns.str.contains('^Unnamed')]

# Adjust data types - wrong types were causing issues
clients.CLIENT_NUMBER = [str(clients) for clients in clients.CLIENT_NUMBER]
cmd.CLIENT_NUMBER = [str(cmd) for cmd in cmd.CLIENT_NUMBER]
table.CLIENT_NUMBER = [str(table) for table in table.CLIENT_NUMBER]
table.VISITOR_ID = [str(table) for table in table.VISITOR_ID]
log.VISITOR_ID = [str(log) for log in log.VISITOR_ID]

# Cannot distinguish between multiple orders on the same day, so the table has been simplified.
cmd = cmd.groupby(['CLIENT_NUMBER', 'ORDER_DATE', 'CVIC']).sum()
# Convert back to typical DataFrame type
cmd = cmd.reset_index()

# Join cmd + table + log

In [374]:
cmd.CLIENT_NUMBER = [str(cmd) for cmd in cmd.CLIENT_NUMBER]
log['EVENT_SHORT'] = log.EVENT_DATE.str.slice(0,10)
data = log[log.VISITOR_ID != '0']
data = pd.merge(data, table, left_on='VISITOR_ID', right_on='VISITOR_ID', how='left')
data = pd.merge(data, cmd, left_on=['CLIENT_NUMBER', 'EVENT_SHORT'], right_on=['CLIENT_NUMBER', 'ORDER_DATE'], how='left')

In [376]:
data

VISITOR_ID CONNECTED_SESSION  ID_SESSION   SESSION_START_DATE  \
0          381225               OUI           1  2017-08-08 00:02:06   
1          563053               OUI          13  2017-08-08 00:01:32   
2          563053               OUI          13  2017-08-08 00:01:32   
3          706588               OUI          23  2017-08-08 00:01:29   
4          572702               OUI          64  2017-08-08 00:03:19   
5          572702               OUI          64  2017-08-08 00:03:19   
6          572702               OUI          64  2017-08-08 00:03:19   
7          698042               OUI         201  2017-08-08 00:14:41   
8          489527               OUI         203  2017-08-08 00:15:00   
9          489527               OUI         203  2017-08-08 00:15:00   
10         528544               OUI         232  2017-08-08 00:17:31   
11         706590               OUI         281  2017-08-08 00:22:35   
12         706590               OUI         281  2017-08-08 00:22:35   
13         484224               OUI         386  2017-08-08 00:33:41   
14         484224               OUI         386  2017-08-08 00:33:41   
15         484224               OUI         386  2017-08-08 00:33:41   
16         498172               OUI         444  2017-08-08 00:40:42   
17         706592               OUI         485  2017-08-08 00:45:40   
18         464818               OUI         517  2017-08-08 00:50:55   
19         464818               OUI         517  2017-08-08 00:50:55   
20         706591               OUI         536  2017-08-08 00:53:45   
21         706294               OUI         572  2017-08-08 01:01:15   
22         685876               OUI         589  2017-08-08 01:05:27   
23         685876               OUI         589  2017-08-08 01:05:27   
24         284469               OUI         611  2017-08-08 01:10:13   
25         284469               OUI         611  2017-08-08 01:10:13   
26         284469               OUI         611  2017-08-08 01:10:13   
27         623112               OUI         660  2017-08-08 01:22:12   
28         623112               OUI         660  2017-08-08 01:22:12   
29         623112               OUI         660  2017-08-08 01:22:12   
...           ...               ...         ...                  ...   
666704     632825               OUI       20287  2017-08-29 14:55:37   
666705     286176               OUI       20466  2017-08-29 15:00:20   
666706     286176               OUI       20466  2017-08-29 15:00:20   
666707     286176               OUI       20466  2017-08-29 15:00:20   
666708     286176               OUI       20466  2017-08-29 15:00:20   
666709     286176               OUI       20466  2017-08-29 15:00:20   
666710     711788               OUI       20705  2017-08-29 15:08:38   
666711     261647               OUI       23579  2017-08-29 16:47:27   
666712     295204               OUI       27180  2017-08-29 18:51:35   
666713     487207               OUI       27852  2017-08-29 19:20:31   
666714     296743               OUI        3950  2017-08-30 08:35:18   
666715     469453               OUI        8477  2017-08-30 09:27:52   
666716     469453               OUI        8477  2017-08-30 09:27:52   
666717     712067               OUI       13963  2017-08-30 11:06:54   
666718     712068               OUI       14326  2017-08-30 11:14:58   
666719     712073               OUI       15349  2017-08-30 11:38:55   
666720     340126               OUI       20535  2017-08-30 14:13:32   
666721     712213               OUI       23158  2017-08-30 14:42:31   
666722     163969               OUI       25495  2017-08-30 15:17:41   
666723     163969               OUI       25495  2017-08-30 15:17:41   
666724     688670               OUI       29342  2017-08-30 16:31:00   
666725     320981               OUI       43593  2017-08-30 23:05:20   
666726     691621               OUI        5839  2017-08-31 08:56:12   
666727     691621               OUI        5839

# Visitor Journey

Looking at the logs of **August 2017** <br>
**Visitor 261647** first connects at **11:25 AM**
<br> <br>
Connected through **Sponsored Links**, on a **Computer** <br>
**19 minutes:** Address entered for delivery.  <br>
**22 minutes:** Accessed via phone to view account <br>
**28 minutes:** Puchase complete.<br>
**28 minutes:** Disconnected from website
<br> <br>
**187.50 Euro Spent**

# What do I want to classify?
If a customer was convinced to purchase through magazine or online.

# Useful Features
- PAGES: Are they browsing the wares or going straight to the purchase?<br>
- GLOBAL_SOURCES: What caused them to connect to the website? <br>
- DEVICE_TYPE: Maybe magazine users use different devices <br>
- CVIC - Did they use a magazine coupon <br>
- PRE_TAX_AMOUNT <br>

# Feature Engineering

- time connected <br>
- number of unique pages visited
- T/F: Placed order 

# Generating the Data Matrix
- VISITOR_ID <br>
- CLIENT_NUMBER <br>
- EVENT_SHORT <br>
- GLOBAL_SOURCES <br>
- DEVICE TYPE <br>
- CVIC <br>
- PRE_TAX_AMOUNT <br>


In [419]:
tmp = data[['VISITOR_ID', 'CLIENT_NUMBER', 'EVENT_SHORT', 'GLOBAL_SOURCES', 'DEVICE_TYPE', 'CVIC', 'PRE_TAX_AMOUNT', 'EVENT_DATE', 'PAGES']].copy()
tmp = tmp.fillna(0)
tmp['PURCHASE'] = True
tmp.loc[tmp.PRE_TAX_AMOUNT == 0, 'PURCHASE'] = False
tmp.columns = ['VISITOR_ID', 'CLIENT_NUMBER', 'DATE', 'SOURCE', 'DEVICE', 'PROMO', 'EXPENSE', 'EVENT_DATE', 'PAGES', 'PURCHASE']
tmp = tmp[['VISITOR_ID', 'CLIENT_NUMBER', 'DATE', 'SOURCE', 'DEVICE', 'PROMO', 'EXPENSE', 'PURCHASE', 'EVENT_DATE', 'PAGES']]

In [433]:
len(set(tmp.CLIENT_NUMBER))

22451

In [421]:
tmp.head()

VISITOR_ID         CLIENT_NUMBER        DATE             SOURCE      DEVICE  \
0     381225   2912426551072989703  2017-08-08  Liens Sponsorisés  Smartphone   
1     563053  -5859629033836166974  2017-08-08       Accès Direct  Ordinateur   
2     563053  -5859629033836166974  2017-08-08       Accès Direct  Ordinateur   
3     706588  -7174350147193854740  2017-08-08       Accès Direct  Ordinateur   
4     572702    866042539231455965  2017-08-08            Moteurs    Tablette   

   PROMO  EXPENSE  PURCHASE  
0      0     0.00     False  
1  False   280.34      True  
2  False   280.34      True  
3  False    19.08      True  
4   True    97.50      True

# Testing

In [431]:
tmp[tmp.PURCHASE == True]

In [432]:
tmp

VISITOR_ID         CLIENT_NUMBER        DATE             SOURCE  \
0          381225   2912426551072989703  2017-08-08  Liens Sponsorisés   
1          563053  -5859629033836166974  2017-08-08       Accès Direct   
2          563053  -5859629033836166974  2017-08-08       Accès Direct   
3          706588  -7174350147193854740  2017-08-08       Accès Direct   
4          572702    866042539231455965  2017-08-08            Moteurs   
5          572702    866042539231455965  2017-08-08            Moteurs   
6          572702    866042539231455965  2017-08-08            Moteurs   
7          698042   3565018203965887144  2017-08-08  Liens Sponsorisés   
8          489527   5954669164526364513  2017-08-08  Liens Sponsorisés   
9          489527   5954669164526364513  2017-08-08  Liens Sponsorisés   
10         528544  -8974800123977524097  2017-08-08    Email marketing   
11         706590     90227983180710098  2017-08-08            Moteurs   
12         706590     90227983180710098  2017-08-08            Moteurs   
13         484224  -7646437119462479870  2017-08-08           Webmails   
14         484224  -7646437119462479870  2017-08-08           Webmails   
15         484224  -7646437119462479870  2017-08-08           Webmails   
16         498172   3460910494336530630  2017-08-08  Liens Sponsorisés   
17         706592  -7471776745553281864  2017-08-08       Accès Direct   
18         464818   5379022874790056044  2017-08-08            Moteurs   
19         464818   5379022874790056044  2017-08-08            Moteurs   
20         706591  -9172043205706441743  2017-08-08  Liens Sponsorisés   
21         706294  -8869054033881510891  2017-08-08            Moteurs   
22         685876   3880709424713964820  2017-08-08       Accès Direct   
23         685876   3880709424713964820  2017-08-08       Accès Direct   
24         284469  -3927683316740594505  2017-08-08            Moteurs   
25         284469  -3927683316740594505  2017-08-08            Moteurs   
26         284469  -3927683316740594505  2017-08-08            Moteurs   
27         623112  -6422665475109442159  2017-08-08    Email marketing   
28         623112  -6422665475109442159  2017-08-08    Email marketing   
29         623112  -6422665475109442159  2017-08-08    Email marketing   
...           ...                   ...         ...                ...   
666704     632825   8810417041213755608  2017-08-29  Liens Sponsorisés   
666705     286176   3253969207395396299  2017-08-29            Moteurs   
666706     286176   3253969207395396299  2017-08-29            Moteurs   
666707     286176   3253969207395396299  2017-08-29            Moteurs   
666708     286176   3253969207395396299  2017-08-29            Moteurs   
666709     286176   3253969207395396299  2017-08-29            Moteurs   
666710     711788   4866961083522369159  2017-08-29  Liens Sponsorisés   
666711     261647  -3951508623707535255  2017-08-29            Moteurs   
666712     295204  -7389978521523108019  2017-08-29            Moteurs   
666713     487207  -4082759471703855136  2017-08-29       Accès Direct   
666714     296743  -1516722727970022212  2017-08-30            Moteurs   
666715     469453   3462776611095962678  2017-08-30  Liens Sponsorisés   
666716     469453   3462776611095962678  2017-08-30  Liens Sponsorisés   
666717     712067  -6669675012336252407  2017-08-30            Moteurs   
666718     712068   8161863904893967570  2017-08-30            Moteurs   
666719     712073   8367580684820619746  2017-08-30            Moteurs   
666720     340126  -7414155220859863662  2017-08-30  Liens Sponsorisés   
666721     712213  -6776147515075359023  2017-08-30            Moteurs   
666722     163969   3981969298899529751  2017-08-30  Liens Sponsorisés   
666723     163969   3981969298899529751  2017-08-30  Liens Sponsorisés   
666724     688670  -2745035742082786583  2017-08-30  Liens Sponsorisés   
666725     320981  -6708892950396807138  2017-08-30  Liens Sponsorisés   
666726     

# VISITOR JOURNEY

In [230]:
v_17 = log[log.VISITOR_ID == '175752']
v_17.head()

VISITOR_ID CONNECTED_SESSION  ID_SESSION   SESSION_START_DATE  \
159645     175752               OUI       28914  2017-08-24 13:27:43   
159646     175752               OUI       28914  2017-08-24 13:27:43   
159647     175752               OUI       28914  2017-08-24 13:27:43   
182418     175752               OUI       14817  2017-08-27 10:17:25   
184863     175752               OUI       33484  2017-08-27 14:46:12   

                 EVENT_DATE    PAGES   GLOBAL_SOURCES DEVICE_TYPE  \
159645  2017-08-24 13:27:43  accueil  Email marketing  Ordinateur   
159646  2017-08-24 13:28:14  accueil  Email marketing  Ordinateur   
159647  2017-08-24 13:41:45  accueil  Email marketing  Ordinateur   
182418  2017-08-27 10:17:25  accueil     Accès Direct  Ordinateur   
184863  2017-08-27 14:46:12  accueil     Accès Direct  Ordinateur   

       DEVICE_MODEL  LOADINGS  
159645          NaN         1  
159646          NaN         1  
159647          NaN         1  
182418          NaN         1  
184863          NaN         1

In [235]:
v_17.sort_values(by=['EVENT_DATE'])

VISITOR_ID CONNECTED_SESSION  ID_SESSION   SESSION_START_DATE  \
159645      175752               OUI       28914  2017-08-24 13:27:43   
2091135     175752               OUI       28914  2017-08-24 13:27:43   
159646      175752               OUI       28914  2017-08-24 13:27:43   
278822      175752               OUI       28914  2017-08-24 13:27:43   
334433      175752               OUI       28914  2017-08-24 13:27:43   
667488      175752               OUI       28914  2017-08-24 13:27:43   
445710      175752               OUI       28914  2017-08-24 13:27:43   
445711      175752               OUI       28914  2017-08-24 13:27:43   
607990      175752               OUI       28914  2017-08-24 13:27:43   
547217      175752               OUI       28914  2017-08-24 13:27:43   
547218      175752               OUI       28914  2017-08-24 13:27:43   
565337      175752               OUI       28914  2017-08-24 13:27:43   
580843      175752               OUI       28914  2017-08-24 13:27:43   
594242      175752               OUI       28914  2017-08-24 13:27:43   
159647      175752               OUI       28914  2017-08-24 13:27:43   
182418      175752               OUI       14817  2017-08-27 10:17:25   
817674      175752               OUI       14817  2017-08-27 10:17:25   
524076      175752               OUI       14817  2017-08-27 10:17:25   
935047      175752               OUI       14817  2017-08-27 10:17:25   
1272059     175752               OUI       14817  2017-08-27 10:17:25   
346755      175752               OUI       14817  2017-08-27 10:17:25   
1272060     175752               OUI       14817  2017-08-27 10:17:25   
1331541     175752               OUI       14817  2017-08-27 10:17:25   
1272061     175752               OUI       14817  2017-08-27 10:17:25   
1193776     175752               OUI       14817  2017-08-27 10:17:25   
1272062     175752               OUI       14817  2017-08-27 10:17:25   
1616557     175752               OUI       14817  2017-08-27 10:17:25   
1272063     175752               OUI       14817  2017-08-27 10:17:25   
1498880     175752               OUI       14817  2017-08-27 10:17:25   
1272064     175752               OUI       14817  2017-08-27 10:17:25   
...            ...               ...         ...                  ...   
1181330     175752               OUI       12691  2017-08-29 10:55:54   
221826      175752               OUI       18234  2017-08-31 10:36:37   
822040      175752               OUI       18234  2017-08-31 10:36:37   
531102      175752               OUI       18234  2017-08-31 10:36:37   
841503      175752               OUI       18234  2017-08-31 10:36:37   
1276809     175752               OUI       18234  2017-08-31 10:36:37   
1616725     175752               OUI       18234  2017-08-31 10:36:37   
1276810     175752               OUI       18234  2017-08-31 10:36:37   
1617779     175752               OUI       18234  2017-08-31 10:36:37   
1276811     175752               OUI       18234  2017-08-31 10:36:37   
1616726     175752               OUI       18234  2017-08-31 10:36:37   
221827      175752               OUI       18234  2017-08-31 10:36:37   
225057      175752               OUI       33524  2017-08-31 14:57:38   
822382      175752               OUI       33524  2017-08-31 14:57:38   
531636      175752               OUI       33524  2017-08-31 14:57:38   
841523      175752               OUI       33524  2017-08-31 14:57:38   
691390      175752               OUI       33524  2017-08-31 14:57:38   
841524      175752               OUI       33524  2017-08-31 14:57:38   
691391      175752               OUI       33524  2017-08-31 14:57:38   
841525      175752               OUI       33524  2017-08-31 14:57:38   
489750      175752               OUI       33524  2017-08-31 14:57:38   
555107      175752               OUI       33524  2017-08-31 14:57:38   
555108      175752               OUI       33524  2017-08-31 14:57